### PR TITLE
[MM-61938] Strip IPv6 Zone from proxied URLs

### PR DESCRIPTION
**Pull request by streamer45**
#### Summary

We implement a proxy function wrapper for our HTTP client implementation that strips IPv6 zone IDs in case they are present since they are causing some issues (see ticket for details).

#### Ticket Link

https://mattermost.atlassian.net/browse/MM-61938

#### Release Note

```release-note
Fixed an issue when making proxied HTTP requests in which the URL had an IPv6 address containing a zone ID.
```

### CODE REVIEW

File: server/public/shared/httpservice/client.go [hunk 1]
```diff
@@ -155,6 +159,26 @@ func dialContextFilter(dial DialContextFunction, allowHost func(host string) boo
 	}
 }
 
+func getProxyFn() func(r *http.Request) (*url.URL, error) {
+	proxyFromEnvFn := httpproxy.FromEnvironment().ProxyFunc()
+	return func(r *http.Request) (*url.URL, error) {
+		// Mitigating against MM-61938
+		// TODO: Remove this code once the above issue is fixed upstream.
+		if r.URL != nil {
+			if addr, err := netip.ParseAddr(r.URL.Hostname()); err == nil && addr.Is6() && addr.Zone() != "" {
```
Comment (agnivade) [inline server/public/shared/httpservice/client.go:168]: Do we need to log a warning if `err != nil`?

File: server/public/shared/httpservice/client.go [hunk 2]
```diff
@@ -155,6 +159,26 @@ func dialContextFilter(dial DialContextFunction, allowHost func(host string) boo
...
+				// Strip IPv6 zone and rebuild URL
+				if r.URL.Port() != "" {
+					r.URL.Host = addr.WithZone("").String() + ":" + r.URL.Port()
+				} else {
+					r.URL.Host = addr.WithZone("").String()
+				}
```
Comment (agnivade) [inline server/public/shared/httpservice/client.go:174]: Might simplify to:

```go
portSuffix := ""
if r.URL.Port() != "" {
	portSuffix = ":" + r.URL.Port()
}
r.URL.Host = addr.WithZone("").String() + portSuffix
```

File: server/public/shared/httpservice/client_test.go [hunk 3]
```diff
@@ -261,3 +261,82 @@ func TestSplitHostnames(t *testing.T) {
 	hostnames = strings.FieldsFunc(config, splitFields)
 	require.Equal(t, []string{"127.0.0.1", "localhost", "192.168.1.0"}, hostnames)
 }
+
+func TestGetProxyFn(t *testing.T) {
+	t.Setenv("HTTP_PROXY", "http://proxy.example.org")
+	t.Setenv("HTTPS_PROXY", "https://proxy.example.org")
+	t.Setenv("NO_PROXY", ".example.com")
```
Comment (agnivade) [inline server/public/shared/httpservice/client_test.go:268]: Oh hey, I wasn't aware of this! Nice!

File: server/public/shared/httpservice/client.go (hunk 2 above)
Comment (streamer45) [reply server/public/shared/httpservice/client.go:174]: point of view: still 5 lines + a new variable :p

**Review by agnivade — APPROVED** [review]

**Review by cpoile — APPROVED** [review]
Nice, and good tests.

**Pull request closed without merging**